tools/infrasys/src/root.rs

```diff
@@ -0,0 +1,196 @@
+    filepath: &String,
+) -> Result<()> {
+    match signing_key_config {
+        SigningKeyConfig::file { .. } => (),
+        SigningKeyConfig::kms { key_id, config, .. } => add_keys_kms(
+            &config
+                .as_ref()
+                .context(error::MissingConfig {
+                    missing: "config field for a kms key",
+                })?
+                .available_keys,
+            role,
+            threshold,
+            filepath,
+            key_id,
+        )?,
+        SigningKeyConfig::ssm { .. } => (),
+    }
+    Ok(())
+}
+
+/// Adds KMSKeys to root.json given root or publication type
+/// Input: available-keys (keys to sign with), role (root or publication), threshold for role, filepath for root.JSON,
+/// mutable key_id
+/// Output: in-place edit of root.json and key_id with a valid publication key
+/// (If key-id is populated, it will not change. Otherwise, it will be populated with a key-id of an available key)
+fn add_keys_kms(
+    available_keys: &HashMap<String, String>,
+    role: &KeyRole,
+    threshold: &String,
+    filepath: &String,
+    key_id: &mut Option<String>,
+) -> Result<()> {
+    if (*available_keys).len()
+        < (*threshold)
+            .parse::<usize>()
+            .context(error::ParseInt { what: threshold })?
+    {
+        error::InvalidThreshold {
+            threshold,
+            num_keys: (*available_keys).len(),
+        }
+        .fail()?;
+    }
+    match role {
+        KeyRole::Root => {
+            tuftool!(
+                Region::default().name(),
+                "root set-threshold '{}' root '{}' ",
+                filepath,
+                threshold
+            );
+            for (keyid, region) in available_keys.iter() {
+                tuftool!(
+                    region,
+                    "root add-key '{}' aws-kms:///'{}' --role root",
+                    filepath,
+                    keyid
+                );
+            }
+        }
+        KeyRole::Publication => {
+            tuftool!(
+                Region::default().name(),
+                "root set-threshold '{}' snapshot '{}' ",
+                filepath,
+                threshold
+            );
+            tuftool!(
+                Region::default().name(),
+                "root set-threshold '{}' targets '{}' ",
+                filepath,
+                threshold
+            );
+            tuftool!(
+                Region::default().name(),
+                "root set-threshold '{}' timestamp '{}' ",
+                filepath,
+                threshold
+            );
+            for (keyid, region) in available_keys.iter() {
+                tuftool!(
+                region,
+                "root add-key '{}' aws-kms:///'{}' --role snapshot --role targets --role timestamp",
+                filepath,
+                keyid
+                );
+            }
+
+            // Set key_id using a publication key (if one is not already provided)
```

Comment:
We must set key_id in this method as it's the only one that differentiates roles (we only want key_id to be set for publication keys, not root keys)